src/stateProviders/provideThing.ts

```diff
@@ -0,0 +1,112 @@
+import _ from 'lodash';
+import { provideState } from 'freactal';
+
+import RESOURCE_MAP from 'common/RESOURCE_MAP';
+
+const provideThing = provideState({
+  initialState: () => ({ item: null, staged: null, associated: {} }),
+
+  effects: {
+    getState: () => state => ({ ...state }),
+    setItem: async (effects, id, type) => {
+      const [item, ...associated] = id
+        ? await Promise.all([
+            RESOURCE_MAP[type].getItem(id),
+            ...RESOURCE_MAP[type].associatedTypes.map(associatedType =>
+              RESOURCE_MAP[associatedType].getList({ [`${type}Id`]: id }),
+            ),
+          ])
+        : [null, ...RESOURCE_MAP[type].associatedTypes.map(() => null)];
+
+      return s => {
+        return {
+          ...s,
+          type,
+          item,
+          id,
+          staged: item,
+          associated: associated.reduce(
+            (acc, a, i) => ({
+              ...acc,
+              [RESOURCE_MAP[type].associatedTypes[i]]: a,
+            }),
+            {},
+          ),
+        };
+      };
+    },
+    stageChange: async (effects, change) => {
+      return state => {
+        return {
+          ...state,
+          staged: { ...state.staged, ..._.omit(change, RESOURCE_MAP[state.type].associatedTypes) },
+          associated: Object.keys(state.associated).reduce((acc, currentType) => {
+            if (change[currentType]) {
+              return {
+                ...acc,
+                [currentType]: {
+                  ...state.associated[currentType],
+                  ...Object.keys(change[currentType]).reduce((acc, action) => {
+                    const otherAction = action === 'add' ? 'remove' : 'add';
+                    if (
+                      (state.associated[currentType][otherAction] || []).includes(
+                        change[currentType][action],
+                      )
+                    ) {
+                      return {
+                        ...acc,
+                        [otherAction]: (state.associated[currentType][otherAction] || []).filter(
+                          ({ id }) => id !== change[currentType][action].id,
+                        ),
+                      };
+                    } else {
+                      return {
+                        ...acc,
+                        [action]: _.uniq([
+                          ...(state.associated[currentType][action] || []),
+                          change[currentType][action],
+                        ]),
+                      };
+                    }
+                  }, {}),
+                },
+              };
+            } else {
+              return { ...acc, [currentType]: state.associated[currentType] };
+            }
+          }, {}),
+        };
+      };
+    },
```

Comment:
L44 to L80 is a bit hard on the eyes ._.
Let's talk about this a bit

Reply:
I think it's definitely mergeable though, but maybe we can try something else after